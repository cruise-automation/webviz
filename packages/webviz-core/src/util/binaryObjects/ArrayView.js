// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

type GetArrayElement<T> = (offset: number) => T;

export type ArrayView<T> = $ReadOnly<{
  get: (index: number) => T,
  length: () => number,
  iter: () => Iterable<T>,
}>;

// Class is inside a closure to make instance construction cheaper (only two fields to set). The
// getElement and elementSize fields are common to many instances.
const getArrayView = <T>(getElement: GetArrayElement<T>, elementSize: number) =>
  class {
    _begin: number;
    _end: number;
    constructor(begin: number, end: number) {
      this._begin = begin;
      this._end = end;
    }

    // Unfortunately we can't override the [] operator without a proxy, which is very slow.
    get(index: number): T {
      return getElement(this._begin + index * elementSize);
    }

    length(): number {
      return (this._end - this._begin) / elementSize;
    }

    *iter(): Iterable<T> {
      for (let i = this._begin; i < this._end; i += elementSize) {
        yield getElement(i);
      }
    }
  };

export default getArrayView;
