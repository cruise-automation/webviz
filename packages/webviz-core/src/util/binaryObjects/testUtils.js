// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import type { RosValue } from "webviz-core/src/players/types";
import type { BinaryHeader } from "webviz-core/src/types/BinaryMessages";
import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";
import type { ArrayView } from "webviz-core/src/util/binaryObjects/ArrayViews";

export const definitions: RosDatatypes = {
  "std_msgs/Header": {
    fields: [{ type: "uint32", name: "seq" }, { type: "time", name: "stamp" }, { type: "string", name: "frame_id" }],
  },
  "fake_msgs/HasComplexAndArray": {
    fields: [{ type: "std_msgs/Header", name: "header" }, { type: "string", isArray: true, name: "stringArray" }],
  },
  "fake_msgs/HasComplexArray": {
    fields: [{ type: "fake_msgs/HasComplexAndArray", name: "complexArray", isArray: true }],
  },
  "fake_msgs/HasConstant": {
    fields: [{ type: "uint8", name: "const", isConstant: true, value: 1 }],
  },
  "fake_msgs/HasByteArray": {
    fields: [{ type: "uint8", name: "byte_array", isArray: true }],
  },
  "fake_msgs/HasJson": {
    fields: [{ type: "json", name: "jsonField" }],
  },
  "fake_msgs/HasInt64s": {
    fields: [{ type: "int64", name: "i64" }, { type: "uint64", name: "u64" }],
  },
  "fake_msgs/HasArrayOfEmpties": {
    fields: [{ type: "fake_msgs/HasConstant", name: "arr", isArray: true }],
  },
  "fake_msgs/HasBigIntArrays": {
    fields: [{ type: "int64", name: "i_arr", isArray: true }, { type: "uint64", name: "u_arr", isArray: true }],
  },
  "fake_msgs/ContainsEverything": {
    fields: [
      { type: "std_msgs/Header", name: "first" },
      { type: "fake_msgs/HasComplexAndArray", name: "second" },
      { type: "fake_msgs/HasComplexArray", name: "third" },
      { type: "fake_msgs/HasConstant", name: "fourth" },
      { type: "fake_msgs/HasByteArray", name: "fifth" },
      { type: "fake_msgs/HasJson", name: "sixth" },
      { type: "fake_msgs/HasInt64s", name: "seventh" },
      { type: "fake_msgs/HasArrayOfEmpties", name: "eighth" },
      { type: "fake_msgs/HasBigIntArrays", name: "ninth" },
    ],
  },
};

export type HasComplexAndArray = $ReadOnly<{|
  header(): BinaryHeader,
  stringArray(): ArrayView<string>,
|}>;

export type HasComplexArray = $ReadOnly<{|
  complexArray(): ArrayView<HasComplexAndArray>,
|}>;

export type HasConstant = $ReadOnly<{|
  const(): number,
|}>;

export type HasByteArray = $ReadOnly<{|
  byte_array(): Uint8Array,
|}>;

export type HasJson = $ReadOnly<{|
  jsonField(): RosValue,
|}>;

export type HasInt64s = $ReadOnly<{|
  i64(bigInt: ?true): number,
  u64(bigInt: ?true): number,
|}>;

export type HasArrayOfEmpties = $ReadOnly<{|
  arr(): ArrayView<HasConstant>,
|}>;

export type HasBigIntArrays = $ReadOnly<{|
  i_arr(): ArrayView<number>,
  u_arr(): ArrayView<number>,
|}>;

export type ContainsEverything = $ReadOnly<{|
  first(): BinaryHeader,
  second(): HasComplexAndArray,
  third(): HasComplexArray,
  fourth(): HasConstant,
  fifth(): HasByteArray,
  sixth(): HasJson,
  seventh(): HasInt64s,
  eighth(): HasArrayOfEmpties,
  ninth(): HasBigIntArrays,
|}>;
