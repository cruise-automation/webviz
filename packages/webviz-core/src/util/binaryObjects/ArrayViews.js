// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { find } from "lodash";

import { deepParse, isBobject } from "webviz-core/src/util/binaryObjects";
import { deepParseSymbol } from "webviz-core/src/util/binaryObjects/messageDefinitionUtils";

type GetArrayElement<T> = (offset: number) => T;

export interface ArrayView<T> {
  +get: (index: number, bigInt: ?true) => T;
  +length: () => number;
  +toArray: () => T[];
  +find: (predicate: (item: T, index: number, collection: T[]) => boolean) => ?T;
  @@iterator(): Iterator<T>;
}

// Class is inside a closure to make instance construction cheaper (only two fields to set). The
// getElement and elementSize fields are common to many instances.
export const getArrayView = <T>(
  getElement: GetArrayElement<T>,
  elementSize: number,
  getBigIntElement: ?GetArrayElement<T>
) =>
  class BinaryArrayView implements ArrayView<T> {
    _begin: number;
    _length: number;
    constructor(begin: number, length: number) {
      this._begin = begin;
      this._length = length;
    }

    // Unfortunately we can't override the [] operator without a proxy, which is very slow.
    get(index: number, bigInt: ?true): T {
      if (bigInt && getBigIntElement) {
        return getBigIntElement(this._begin + index * elementSize);
      }
      return getElement(this._begin + index * elementSize);
    }

    length(): number {
      return this._length;
    }

    // https://stackoverflow.com/questions/48491307/iterable-class-in-flow
    /*::
    @@iterator(): Iterator<T> {
      // $FlowFixMe
      return this[Symbol.iterator]()
    }
    */
    // $FlowFixMe
    *[Symbol.iterator](): Iterable<T> {
      let offset = this._begin;
      const length = this._length;
      for (let i = 0; i < length; i += 1) {
        yield getElement(offset);
        offset += elementSize;
      }
    }

    // $FlowFixMe
    [deepParseSymbol](): T[] {
      const ret = new Array(this._length);
      let offset = this._begin;
      const length = this._length;
      for (let i = 0; i < length; i += 1) {
        const o = getElement(offset);
        ret[i] = isBobject(o) ? deepParse(o) : o;
        offset += elementSize;
      }
      return ret;
    }

    // Shallow parse. Equivalent to [...this], but faster. Used in deep parsing for primitive
    // types, so quite performance-sensitive.
    toArray(): T[] {
      const ret = new Array(this.length());
      let offset = this._begin;
      const length = this._length;
      for (let i = 0; i < length; i += 1) {
        ret[i] = getElement(offset);
        offset += elementSize;
      }
      return ret;
    }

    find(predicate: (item: T, index: number, collection: T[]) => boolean): ?T {
      return find(this.toArray(), predicate);
    }

    offset(): number {
      // It would be nice to just define an [offsetSymbol] attribute, but flow really dislikes it.
      return this._begin;
    }
  };

export class PrimitiveArrayView<T> implements ArrayView<T> {
  value: T[];
  constructor(value: T[]) {
    this.value = value;
  }
  get(index: number): T {
    return this.value[index];
  }
  length() {
    return this.value.length;
  }
  // https://stackoverflow.com/questions/48491307/iterable-class-in-flow
  /*::
  @@iterator(): Iterator<T> {
    // $FlowFixMe
    return this[Symbol.iterator]()
  }
  */
  // $FlowFixMe
  *[Symbol.iterator](): Iterable<T> {
    for (const o of this.value) {
      yield o;
    }
  }
  // Use deepParse(arr)
  // $FlowFixMe
  [deepParseSymbol](): T[] {
    return this.value;
  }
  toArray(): T[] {
    return this.value;
  }

  find(predicate: (item: T, index: number, collection: T[]) => boolean): ?T {
    return find(this.toArray(), predicate);
  }
}

export const getReverseWrapperArrayView = <T>(Class: any) =>
  class ReverseWrapperArrayView implements ArrayView<T> {
    value: T[];
    constructor(value: T[]) {
      this.value = value;
    }
    get(index: number): T {
      return new Class(this.value[index]);
    }
    length(): number {
      return this.value.length;
    }
    // https://stackoverflow.com/questions/48491307/iterable-class-in-flow
    /*::
    @@iterator(): Iterator<T> {
      // $FlowFixMe
      return this[Symbol.iterator]()
    }
    */
    // $FlowFixMe
    *[Symbol.iterator](): Iterable<T> {
      for (const o of this.value) {
        yield isBobject(o) ? o : new Class(o);
      }
    }
    // Use deepParse(arr)
    // $FlowFixMe
    [deepParseSymbol](): T[] {
      return this.value.map((o) => (isBobject(o) ? deepParse(o) : deepParse(new Class(o))));
    }
    toArray(): T[] {
      const ret = [];
      let i = 0;
      for (const o of this) {
        ret[i] = o;
        i += 1;
      }
      return ret;
    }

    find(predicate: (item: T, index: number, collection: T[]) => boolean): ?T {
      return find(this.toArray(), predicate);
    }
  };
