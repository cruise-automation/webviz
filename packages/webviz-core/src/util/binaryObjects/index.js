// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { Buffer } from "buffer";
import int53 from "int53";

import { cast, type Bobject } from "webviz-core/src/players/types";
import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";
import { type ArrayView, getArrayView } from "webviz-core/src/util/binaryObjects/ArrayViews";
import getGetClassForView from "webviz-core/src/util/binaryObjects/binaryWrapperObjects";
import getJsWrapperClasses from "webviz-core/src/util/binaryObjects/jsWrapperObjects";
import {
  associateSourceData,
  deepParseSymbol,
  getSourceData,
  primitiveList,
} from "webviz-core/src/util/binaryObjects/messageDefinitionUtils";

const parseJson = (s) => {
  try {
    return JSON.parse(s);
  } catch (e) {
    return `Could not parse ${JSON.stringify(s)}`;
  }
};
const context = {
  Buffer,
  getArrayView,
  deepParse: deepParseSymbol,
  int53,
  associateSourceData,
  parseJson,
  offsetSymbol: Symbol(),
};

export type { ArrayView };

// Stored for top-level bobjects only.
const approximateBinarySizes = new WeakMap<any, number>();
const reverseWrappedBobjects = new WeakSet<any>();

export const getBinaryOffset = (obj: any): number => {
  const offset = obj[context.offsetSymbol];
  if (offset == null) {
    throw new Error("Can't get offset of non-binary bobject.");
  }
  return offset;
};

export const getBinaryArrayView = (
  typesByName: RosDatatypes,
  datatype: string,
  buffer: ArrayBuffer,
  bigString: string,
  firstElementOffset: number,
  length: number
): ArrayView<any> => {
  const Class = getGetClassForView(typesByName, datatype, true)(context, new DataView(buffer), bigString, typesByName);
  return new Class(firstElementOffset, length);
};

export const getObject = (
  typesByName: RosDatatypes,
  datatype: string,
  buffer: ArrayBuffer,
  bigString: string
): Bobject => {
  const Class = getGetClassForView(typesByName, datatype)(context, new DataView(buffer), bigString, typesByName);
  const ret = new Class(0);
  approximateBinarySizes.set(ret, buffer.byteLength + bigString.length);
  return ret;
};

export const getObjects = (
  typesByName: RosDatatypes,
  datatype: string,
  buffer: ArrayBuffer,
  bigString: string,
  offsets: $ReadOnlyArray<number>
): $ReadOnlyArray<Bobject> => {
  const Class = getGetClassForView(typesByName, datatype)(context, new DataView(buffer), bigString, typesByName);
  const ret = offsets.map((offset) => new Class(offset));
  // Super dumb heuristic: assume all of the bobjects in a block have the same size. We could do
  // better, but this is only used for memory cache eviction which we do a whole block at a time, so
  // we're only actually interested in the sum (which is correct, plus or minus floating point
  // error).
  const approximateSize = (buffer.byteLength + bigString.length) / offsets.length;
  ret.forEach((bobject) => {
    approximateBinarySizes.set(bobject, approximateSize);
  });
  return ret;
};

// True for "object bobjects" and array views.
export const isBobject = (object: ?any): boolean => object?.[deepParseSymbol] != null;
export const isArrayView = (object: any): boolean => isBobject(object) && object[Symbol.iterator] != null;

export const deepParse = (object: ?any): any => {
  if (object == null) {
    // Missing submessage fields are unfortunately common for constructed markers. This is not
    // principled, but it is pragmatic.
    return object;
  }
  if (!isBobject(object)) {
    // This is typically a typing mistake -- the user thinks they have a bobject but have a
    // primitive or a parsed message.
    throw new Error("Argument to deepParse is not a bobject");
  }
  return object[deepParseSymbol]();
};

export const wrapJsObject = <T>(typesByName: RosDatatypes, typeName: string, object: any): T => {
  if (!primitiveList.has(typeName) && !typesByName[typeName]) {
    throw new Error(`Message definition is not present for type ${typeName}.`);
  }
  const classes = getJsWrapperClasses(typesByName);
  const ret = new classes[typeName](object);
  reverseWrappedBobjects.add(ret);
  return cast<T>(ret);
};

// NOTE: The only guarantee is that the sum of the sizes of the bobjects in a given block are
// about right. Sizes are not available for submessages, only top-level bobjects.
// In the future we might make this accurate by getting the data from the binary rewrite step,
// or we might remove this function and just provide access to the identity of the underlying data
// so the shared storage is clear.
export const inaccurateByteSize = (obj: any): number => {
  const size = approximateBinarySizes.get(obj);
  if (size != null) {
    return size;
  }
  if (reverseWrappedBobjects.has(obj)) {
    // Not ideal: Storing the deep-parsed representation of a reverse-wrapped bobject actually
    // does take up memory -- and quite a bit of it. Unfortunately, we don't have a good heuristic
    // for reverse-wrapped bobject sizes.
    return 0;
  }
  throw new Error("Size of object not available");
};

export function bobjectFieldNames(bobject: {}): string[] {
  const sourceData = getSourceData(Object.getPrototypeOf(bobject).constructor);
  if (!sourceData) {
    throw new Error("Unknown constructor in bobjectFieldNames");
  }
  const { datatypes, datatype } = sourceData;
  const definition = datatypes[datatype];
  if (definition == null) {
    if (datatype === "time" || datatype === "duration") {
      return ["sec", "nsec"];
    }
    throw new Error(`Unknown datatype ${datatype}`);
  }
  return definition.fields.filter(({ isConstant }) => !isConstant).map(({ name }) => name);
}

export const fieldNames = (o: {}): string[] => {
  if (!isBobject(o)) {
    return Object.keys(o);
  }
  return bobjectFieldNames(o);
};

export const merge = <T: {}>(bobject: T, overrides: $ReadOnly<{ [field: string]: any }>): T => {
  if (!isBobject(bobject)) {
    throw new Error("Argument to merge is not a bobject");
  }
  const shallow = {};
  // Iterate over class's methods, except `constructor` which is special.
  bobjectFieldNames(bobject).forEach((field) => {
    shallow[field] = bobject[field]();
  });
  const sourceData = getSourceData(Object.getPrototypeOf(bobject).constructor);
  if (sourceData == null) {
    throw new Error("Unknown type in merge");
  }
  return cast<T>(wrapJsObject(sourceData.datatypes, sourceData.datatype, { ...shallow, ...overrides }));
};

// For accessing fields that might be in bobjects and might be in JS objects.
export const getField = (obj: ?any, field: string, bigInts: ?true): any => {
  if (!obj) {
    return;
  }
  if (isBobject(obj)) {
    return obj[field] && obj[field](bigInts);
  }
  return obj[field];
};

export const getIndex = (obj: any, i: number, bigInt: ?true): ?any => {
  if (!obj) {
    return;
  }
  if (isArrayView(obj)) {
    if (i < 0 || i >= obj.length() || !Number.isInteger(i)) {
      return;
    }
    return obj.get(i, bigInt);
  }
  return obj[i];
};

// Get an individual field by traversing a path of keys and indices
export const getFieldFromPath = (obj: any, path: (string | number)[]): any => {
  let ret = obj;
  for (const field of path) {
    ret = typeof field === "string" ? getField(ret, field) : getIndex(ret, field);
  }
  return ret;
};
