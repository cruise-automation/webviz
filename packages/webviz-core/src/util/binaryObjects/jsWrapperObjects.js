// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import indent from "indent-string";
import memoize from "memoize-weak";
import { type RosMsgField } from "rosbag";

import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";
import { PrimitiveArrayView, getReverseWrapperArrayView } from "webviz-core/src/util/binaryObjects/ArrayViews";
import {
  addTimeTypes,
  associateSourceData,
  deepParseSymbol,
  friendlyTypeName,
  isComplex,
} from "webviz-core/src/util/binaryObjects/messageDefinitionUtils";

const arrayTypeName = (typeName: string): string => `${friendlyTypeName(typeName)}$Array`;

const printFieldDefinitionBody = (field: RosMsgField): string => {
  if (field.isConstant) {
    const value = JSON.stringify(field.value);
    if (value == null) {
      throw new Error(`Could not serialize constant value for field ${field.name}`);
    }
    return `return ${value};`;
  }
  const complexExpression = (type) => `const $fieldValue = this[$value].${field.name};
if ($fieldValue == null || $fieldValue[$deepParse] != null) {
  return $fieldValue;
}
return new ${type}($fieldValue);`;
  if (field.isArray && field.type !== "int8" && field.type !== "uint8") {
    if (isComplex(field.type) || field.type === "time" || field.type === "duration") {
      return complexExpression(arrayTypeName(field.type));
    }
    return complexExpression("$PrimitiveArrayView");
  }
  if (isComplex(field.type) || field.type === "time" || field.type === "duration") {
    return complexExpression(friendlyTypeName(field.type));
  }
  // Primitives and byte arrays -- just return as-is, no bobject or null checks.
  return `return this[$value].${field.name};`;
};

// Exported for tests
export const printFieldDefinition = (field: RosMsgField): string => {
  const body = printFieldDefinitionBody(field);
  return [`${field.name}() {`, indent(body, 2), "}"].join("\n");
};

const deepParseFieldExpression = ({ name, type, isArray }) => {
  const isRealArray = isArray && type !== "int8" && type !== "uint8";
  const isRealComplex = isComplex(type) || type === "time" || type === "duration";
  if (isRealArray || isRealComplex) {
    return `$maybeDeepParse(this.${name}())`;
  }
  // Primitives and byte arrays -- just return as-is, no bobject or null checks.
  return `this[$value].${name}`;
};

const printClassDefinition = (typesByName: RosDatatypes, typeName: string): string => {
  const type = typesByName[typeName];
  if (type == null) {
    throw new Error(`Unknown type "${typeName}"`);
  }
  const fieldDefinitions = type.fields.map((field) => indent(printFieldDefinition(field), 2));

  const deepParseFieldExpressions = type.fields
    .filter(({ isConstant }) => !isConstant)
    .map((field) => `${field.name}: ${deepParseFieldExpression(field)},`);

  return `class ${friendlyTypeName(typeName)} {
  constructor(value) {
    this[$value] = value;
  }
${fieldDefinitions.join("\n")}
  [$deepParse]() {
    return {
${indent(deepParseFieldExpressions.join("\n"), 6)}
    };
  }
}
const ${arrayTypeName(typeName)} = $context.getReverseWrapperArrayView(${friendlyTypeName(typeName)});
`;
};

// Exported for tests
export const printClasses = (inputTypesByName: RosDatatypes): string => {
  const typesByName = addTimeTypes(inputTypesByName);
  const classDefinitions = Object.keys(typesByName).map((typeName) => printClassDefinition(typesByName, typeName));

  const classExpressions = Object.keys(typesByName).map(
    (typeName) => `${JSON.stringify(typeName)}: ${friendlyTypeName(typeName)},`
  );

  // Add "maybe deep parse" because `msg.field()?.[$deepParse]()` isn't supported in node.
  return `const $value = Symbol();
const $deepParse = $context.deepParse;
const $maybeDeepParse = (o) => o && o[$deepParse]()
const $PrimitiveArrayView = $context.PrimitiveArrayView;
${classDefinitions.join("\n")}

return {
${indent(classExpressions.join("\n"), 2)}
};`;
};

const getJsWrapperClasses = memoize(
  (datatypes: RosDatatypes): { [typeName: string]: any } => {
    const context = { deepParse: deepParseSymbol, PrimitiveArrayView, getReverseWrapperArrayView };
    /* eslint-disable no-new-func */
    // $FlowFixMe
    const classes = Function("$context", printClasses(datatypes))(context);
    /* eslint-enable no-new-func */
    Object.keys(classes).forEach((datatype) => {
      associateSourceData(classes[datatype], { datatypes, datatype });
    });
    return classes;
  }
);

export default getJsWrapperClasses;
