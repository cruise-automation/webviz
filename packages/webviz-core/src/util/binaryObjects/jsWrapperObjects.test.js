// @flow
//
//  Copyright (c) 2020-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { definitions } from "./testUtils";
import { printClasses } from "webviz-core/src/util/binaryObjects/jsWrapperObjects";

describe("printClasses", () => {
  it("prints the expected function definition", () => {
    expect(printClasses(definitions)).toMatchSnapshot();
  });
});
