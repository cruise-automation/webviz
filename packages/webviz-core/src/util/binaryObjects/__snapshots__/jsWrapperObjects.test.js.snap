// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printClasses prints the expected function definition 1`] = `
"const $value = Symbol();
const $deepParse = $context.deepParse;
const $maybeDeepParse = (o) => o && o[$deepParse]()
const $PrimitiveArrayView = $context.PrimitiveArrayView;
class std_msgs_Header {
  constructor(value) {
    this[$value] = value;
  }
  seq() {
    return this[$value].seq;
  }
  stamp() {
    const $fieldValue = this[$value].stamp;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new time($fieldValue);
  }
  frame_id() {
    return this[$value].frame_id;
  }
  [$deepParse]() {
    return {
      seq: this[$value].seq,
      stamp: $maybeDeepParse(this.stamp()),
      frame_id: this[$value].frame_id,
    };
  }
}
const std_msgs_Header$Array = $context.getReverseWrapperArrayView(std_msgs_Header);

class fake_msgs_HasComplexAndArray {
  constructor(value) {
    this[$value] = value;
  }
  header() {
    const $fieldValue = this[$value].header;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new std_msgs_Header($fieldValue);
  }
  stringArray() {
    const $fieldValue = this[$value].stringArray;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new $PrimitiveArrayView($fieldValue);
  }
  [$deepParse]() {
    return {
      header: $maybeDeepParse(this.header()),
      stringArray: $maybeDeepParse(this.stringArray()),
    };
  }
}
const fake_msgs_HasComplexAndArray$Array = $context.getReverseWrapperArrayView(fake_msgs_HasComplexAndArray);

class fake_msgs_HasComplexArray {
  constructor(value) {
    this[$value] = value;
  }
  complexArray() {
    const $fieldValue = this[$value].complexArray;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasComplexAndArray$Array($fieldValue);
  }
  [$deepParse]() {
    return {
      complexArray: $maybeDeepParse(this.complexArray()),
    };
  }
}
const fake_msgs_HasComplexArray$Array = $context.getReverseWrapperArrayView(fake_msgs_HasComplexArray);

class fake_msgs_HasConstant {
  constructor(value) {
    this[$value] = value;
  }
  const() {
    return 1;
  }
  [$deepParse]() {
    return {

    };
  }
}
const fake_msgs_HasConstant$Array = $context.getReverseWrapperArrayView(fake_msgs_HasConstant);

class fake_msgs_HasByteArray {
  constructor(value) {
    this[$value] = value;
  }
  byte_array() {
    return this[$value].byte_array;
  }
  [$deepParse]() {
    return {
      byte_array: this[$value].byte_array,
    };
  }
}
const fake_msgs_HasByteArray$Array = $context.getReverseWrapperArrayView(fake_msgs_HasByteArray);

class fake_msgs_HasJson {
  constructor(value) {
    this[$value] = value;
  }
  jsonField() {
    return this[$value].jsonField;
  }
  [$deepParse]() {
    return {
      jsonField: this[$value].jsonField,
    };
  }
}
const fake_msgs_HasJson$Array = $context.getReverseWrapperArrayView(fake_msgs_HasJson);

class fake_msgs_HasInt64s {
  constructor(value) {
    this[$value] = value;
  }
  i64() {
    return this[$value].i64;
  }
  u64() {
    return this[$value].u64;
  }
  [$deepParse]() {
    return {
      i64: this[$value].i64,
      u64: this[$value].u64,
    };
  }
}
const fake_msgs_HasInt64s$Array = $context.getReverseWrapperArrayView(fake_msgs_HasInt64s);

class fake_msgs_HasArrayOfEmpties {
  constructor(value) {
    this[$value] = value;
  }
  arr() {
    const $fieldValue = this[$value].arr;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasConstant$Array($fieldValue);
  }
  [$deepParse]() {
    return {
      arr: $maybeDeepParse(this.arr()),
    };
  }
}
const fake_msgs_HasArrayOfEmpties$Array = $context.getReverseWrapperArrayView(fake_msgs_HasArrayOfEmpties);

class fake_msgs_HasBigIntArrays {
  constructor(value) {
    this[$value] = value;
  }
  i_arr() {
    const $fieldValue = this[$value].i_arr;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new $PrimitiveArrayView($fieldValue);
  }
  u_arr() {
    const $fieldValue = this[$value].u_arr;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new $PrimitiveArrayView($fieldValue);
  }
  [$deepParse]() {
    return {
      i_arr: $maybeDeepParse(this.i_arr()),
      u_arr: $maybeDeepParse(this.u_arr()),
    };
  }
}
const fake_msgs_HasBigIntArrays$Array = $context.getReverseWrapperArrayView(fake_msgs_HasBigIntArrays);

class fake_msgs_ContainsEverything {
  constructor(value) {
    this[$value] = value;
  }
  first() {
    const $fieldValue = this[$value].first;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new std_msgs_Header($fieldValue);
  }
  second() {
    const $fieldValue = this[$value].second;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasComplexAndArray($fieldValue);
  }
  third() {
    const $fieldValue = this[$value].third;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasComplexArray($fieldValue);
  }
  fourth() {
    const $fieldValue = this[$value].fourth;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasConstant($fieldValue);
  }
  fifth() {
    const $fieldValue = this[$value].fifth;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasByteArray($fieldValue);
  }
  sixth() {
    const $fieldValue = this[$value].sixth;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasJson($fieldValue);
  }
  seventh() {
    const $fieldValue = this[$value].seventh;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasInt64s($fieldValue);
  }
  eighth() {
    const $fieldValue = this[$value].eighth;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasArrayOfEmpties($fieldValue);
  }
  ninth() {
    const $fieldValue = this[$value].ninth;
    if ($fieldValue == null || $fieldValue[$deepParse] != null) {
      return $fieldValue;
    }
    return new fake_msgs_HasBigIntArrays($fieldValue);
  }
  [$deepParse]() {
    return {
      first: $maybeDeepParse(this.first()),
      second: $maybeDeepParse(this.second()),
      third: $maybeDeepParse(this.third()),
      fourth: $maybeDeepParse(this.fourth()),
      fifth: $maybeDeepParse(this.fifth()),
      sixth: $maybeDeepParse(this.sixth()),
      seventh: $maybeDeepParse(this.seventh()),
      eighth: $maybeDeepParse(this.eighth()),
      ninth: $maybeDeepParse(this.ninth()),
    };
  }
}
const fake_msgs_ContainsEverything$Array = $context.getReverseWrapperArrayView(fake_msgs_ContainsEverything);

class time {
  constructor(value) {
    this[$value] = value;
  }
  sec() {
    return this[$value].sec;
  }
  nsec() {
    return this[$value].nsec;
  }
  [$deepParse]() {
    return {
      sec: this[$value].sec,
      nsec: this[$value].nsec,
    };
  }
}
const time$Array = $context.getReverseWrapperArrayView(time);

class duration {
  constructor(value) {
    this[$value] = value;
  }
  sec() {
    return this[$value].sec;
  }
  nsec() {
    return this[$value].nsec;
  }
  [$deepParse]() {
    return {
      sec: this[$value].sec,
      nsec: this[$value].nsec,
    };
  }
}
const duration$Array = $context.getReverseWrapperArrayView(duration);


return {
  \\"std_msgs/Header\\": std_msgs_Header,
  \\"fake_msgs/HasComplexAndArray\\": fake_msgs_HasComplexAndArray,
  \\"fake_msgs/HasComplexArray\\": fake_msgs_HasComplexArray,
  \\"fake_msgs/HasConstant\\": fake_msgs_HasConstant,
  \\"fake_msgs/HasByteArray\\": fake_msgs_HasByteArray,
  \\"fake_msgs/HasJson\\": fake_msgs_HasJson,
  \\"fake_msgs/HasInt64s\\": fake_msgs_HasInt64s,
  \\"fake_msgs/HasArrayOfEmpties\\": fake_msgs_HasArrayOfEmpties,
  \\"fake_msgs/HasBigIntArrays\\": fake_msgs_HasBigIntArrays,
  \\"fake_msgs/ContainsEverything\\": fake_msgs_ContainsEverything,
  \\"time\\": time,
  \\"duration\\": duration,
};"
`;
